aten/src/ATen/cuda/CUDAStream.cpp

```diff
@@ -1,57 +1,127 @@
 #include "ATen/Error.h"
 
 #include <mutex>
 #include <atomic>
+#include <cstdint>
+#include <deque>
+#include <vector>
 
 // Internal implementation is entirely hidden
+// Note: CUDAStreamInternals doubles for a THCStream
 struct CUDAStreamInternals {
-  bool is_destructible;
-  std::atomic<int> refcount;
-  int64_t device; // Note: cudaGetDevice works with int32_t, not int64_t
-  cudaStream_t stream;
+  CUDAStreamInternals() = default;
+
+  ~CUDAStreamInternals() {
+    if (stream) cudaStreamDestroy(stream);
+  }
+
+  int64_t device = -1; 
+  cudaStream_t stream = nullptr;
 };
 
 namespace at {
 namespace cuda {
 
 namespace detail {
 
-  /*
-  * Stream state
-  */
-  static constexpr cudaStream_t DEFAULT_STREAM = 0;
-
-  static std::once_flag init_flag;
+  // Global stream state and constants
   static int64_t num_gpus;
-  static CUDAStreamInternals* default_streams;
+  static constexpr int STREAMS_PER_POOL = 32;
+  static constexpr unsigned int DEFAULT_FLAGS = cudaStreamNonBlocking;
+  static int HIGH_PRIORITY = 0;
+  static int LOW_PRIORITY = 0;
+  
+  // Default streams
+  static std::once_flag init_flag;
+  static std::vector<CUDAStreamInternals> default_streams;
+
+  // Non-default streams
+  static std::deque<std::once_flag> device_flags;
+  static std::deque<std::atomic<int>> low_priority_counters;
+  static std::deque<std::atomic<int>> high_priority_counters;
+  static std::vector<std::vector<CUDAStreamInternals>> low_priority_streams;
+  static std::vector<std::vector<CUDAStreamInternals>> high_priority_streams;
+
+  // Thread-local current streams
   static thread_local CUDAStreamInternals** current_streams = nullptr;
 
-  // Creates a(n indestructible) default stream for each device
-  // Note: the default stream on each device is signified by a zero
-  // value for the pointer, and so is not actually created as usual.
+  // Populates global values and creates a default stream for each device.
+  // Note: the default stream on each device is signified by a nullptr, 
+  // and so is not created as usual.
   // In particular, we don't need to switch devices when creating the
   // streams.
-  static void initDefaultCUDAStreams() {
+  static void initGlobalStreamState() {
     num_gpus = getNumGPUs();
-    default_streams = (CUDAStreamInternals*) malloc(num_gpus * sizeof(CUDAStreamInternals));
+
+    // Resizes deques and vectors
+    default_streams.resize(num_gpus);
+    device_flags.resize(num_gpus);
+    low_priority_counters.resize(num_gpus);
+    high_priority_counters.resize(num_gpus);
+    low_priority_streams.resize(num_gpus);
+    high_priority_streams.resize(num_gpus);
+
+    // Initializes streams and counters
     for (auto i = decltype(num_gpus){0}; i < num_gpus; ++i) {
-      default_streams[i].is_destructible = false;
-      default_streams[i].refcount = 0;
       default_streams[i].device = i;
-      default_streams[i].stream = DEFAULT_STREAM;
+      low_priority_counters[i] = 0;
+      high_priority_counters[i] = 0;
```

Comment:
Pro-tip: you can initialize the counters to zero upon resize using `low_priority_counters.resize(num_gpus, 0)`